core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3483,19 +3456,7 @@ class KafkaApis(val requestChannel: RequestChannel,
   }
 
   def handleAllocateProducerIdsRequest(request: RequestChannel.Request): Unit = {
-    val zkSupport = metadataSupport.requireZkOrThrow(KafkaApis.shouldNeverReceive(request))
-    authHelper.authorizeClusterOperation(request, CLUSTER_ACTION)
-
-    val allocateProducerIdsRequest = request.body[AllocateProducerIdsRequest]
-
-    if (!zkSupport.controller.isActive)
-      requestHelper.sendResponseMaybeThrottle(request, throttleTimeMs =>
-        allocateProducerIdsRequest.getErrorResponse(throttleTimeMs, Errors.NOT_CONTROLLER.exception))
-    else
-      zkSupport.controller.allocateProducerIds(allocateProducerIdsRequest.data, producerIdsResponse =>
-        requestHelper.sendResponseMaybeThrottle(request, throttleTimeMs =>
-          new AllocateProducerIdsResponse(producerIdsResponse.setThrottleTimeMs(throttleTimeMs)))
-      )
+    throw KafkaApis.shouldNeverReceive(request)
```

Comment:
I think we shouldn't handle it like this - this is one of the request types that is only handled by the controller listener (now that zk is gone):

> "listeners": ["zkBroker", "controller"]

So, we should have generic logic that handles that automatically before it gets here. We can probably just delete this whole method.

Reply:
Same for `handleAlterClientQuotasRequest`.

Reply:
> So, we should have generic logic that handles that automatically before it gets here. We can probably just delete this whole method.

totally agree that we should have a generic logic for those zk-related handler. The socket server can reject the zk-related requests automatically, and we can rewrite `maybeForwardToController` to throw `KafkaApis.shouldAlwaysForward` automatically. With those change, those zk-related handlers can be deleted directly.

In short, the line `case ApiKeys.ALLOCATE_PRODUCER_IDS => handleAllocateProducerIdsRequest(request)` can be removed.

Reply:
Yes, exactly. We can also remove the zk listener from the json files and the socket server will automatically do the right thing, I believe.